src/tensor_ops/boolean/mod.rs

```diff
@@ -0,0 +1,320 @@
+mod cpu_kernels;
+
+#[cfg(feature = "cuda")]
+mod cuda_kernels;
+
+use crate::{
+    prelude::{OnesTensor, Tensor, ZerosTensor},
+    shapes::*,
+    tensor::DeviceStorage,
+};
+
+use std::ops::{BitAnd, BitOr, BitXor, Not};
+
+use super::Device;
+
+pub trait BooleanKernel: DeviceStorage + OnesTensor<bool> + ZerosTensor<bool> {
+    fn not<S: Shape>(
+        &self,
+        inp: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn and<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn or<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn xor<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: &Self::Storage<S, bool>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+
+    fn scalar_and<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, bool>,
+        rhs: bool,
+    ) -> Result<Self::Storage<S, bool>, Self::Err> {
+        if rhs {
+            Ok(lhs.clone())
+        } else {
+            Ok(self.try_zeros_like(lhs)?.storage)
+        }
```

Comment:
I think all three of these scalar methods can be removed from this trait and implemented directly in the macro below (or separately). There's weirdness in using tensor methods, unwrapping them for the storage, and then re-wrapping them in the traits down below.